Photoemission PR

wuppersaver opened the pull request:
Dear all,
it has been a long time but we finally managed to get all the things working as expected.

This is the final version of the code with tests and all the functionality necessary, that should be able to be merged into the code.

I undid a lot of my previous commits on the develop branch to not have a messed up git timeline. This also unfortunately deleted the old PR of mine.

wuppersaver committed:
Applied photoemission patch file

wuppersaver committed:
removed unused variables and minor update to test setup

wuppersaver committed:
edited github workflow

wuppersaver committed:
changed faulty file

wuppersaver committed:
compiling errors with argument mismatch trial fix

wuppersaver committed:
Updated pre-commit action version

wuppersaver committed:
ran precommit locally

ajm143 commented:
Many thanks for this. I'm currently on shared-parental leave so might be a bit slower than usual. 
All the best,
Andrew

ajm143 commented:
- /src/build.f90 seems to have been committed by mistake. This file is made at compile time.
-  Question: What happens (and what should happen) if someone choses TASK: ALL in the odi.
-  Add yourself + Nic to Contributors in the user guide.
- User Guide: 5.2.1 requires the photoemission task
-  It appears that it requires a `devel_code` in the CASTEP param file to make it work? What's your timeline for having this properly in CASTEP?

wuppersaver commented:
Dear Andrew,

thank you for looking into this! My apologies for getting back to you later, I have been quite busy with family matters recently.

I will submit another commit with the new changes.

- The build.f90 has been removed.
- For now, I have not modified the set of tasks called by 'ALL'. There are a number of photoemission parameters where it is difficult to find sane defaults for (work function, photon energy, mean free path, ... ) and not all the files necessary for the photoemission calculation are currently produced by CASTEP when choosing SPECTRAL_TASK as 'ALL'.
- I have added the people from the project on the second page with a note about photoemission. 
Question: Is that a good place or is there a better one? 
I will also add in the citation for the photoemission module paper once it has been published.
- I have added the task and a short note on the differences between the two modes that are currently implemented.
- I have not yet fully finalised the code for submission to the CASTEP repository, but it is very close, so I hope this can become part of the 26.1 version.

wuppersaver committed:
Removed build.f90, edited parameters and user_guide

wuppersaver committed:
Updates for ease of use
- added UserGuide pdf
- updated test benchmark
- added new layer assignment scheme,
more user input, but more flexible
- moved gkgrid size to file rather than user input,
much easier and less error prone

wuppersaver committed:
Minor updates and renaming
Updated user guide
updated od2od help output
updated naming ebind_ptrans_map to reflect output of kinetic energy and not binding energy

wuppersaver committed:
Updated parameter printing and faster gauss function

wuppersaver committed:
Updates to photoemission module

wuppersaver committed:
Update to documentation
Inserted missing titles into bibtex file
Updated the user guide to reflect latest version of the code
Changed program structure diagram (Fig 4.2) to tikz figure and included photoemission module

wuppersaver committed:
Updated gaussian function
Made the gaussian function into a pure function with slightly altered flow for easier
compiler optimisation and more effificent compute

wuppersaver committed:
Updated constants
Removed no longer necessary constants and updated some

wuppersaver committed:
Brought Photoemission module to latest version
Parameters.f90:
Updated parameter names, parameter printing, error printing,
added extra logic for default values and checking input file
Photo.f90:
Added dos dependent model by Saha, et al.
Reworked field emission subroutine due to unit conversion issues and faulty emission probability estimates
Schottky Nordheim emission is now calculated using numerical integration
with recursive bisection and Simpson's rule
Removed old delta function approach, which neglected partially occupied bands
Updated full_momentum_tensor to use all bands during calculation
Updated extra map and curve functions to use already calculated QE values from main QE functions to save on time

wuppersaver committed:
Added large testing set for photoemission module
Updated Cu100 5L slab checkpoint
Updated photoemission testing to 30+ individual tests
Added testing matrix to jobconfig
Edited and added parsers for use in tests

wuppersaver committed:
Resolved test benchmark mismatch

wuppersaver committed:
Upgrade GitHub Actions to v3 for checkout and setup-python

wuppersaver committed:
Upgrade GitHub actions to version 6

wuppersaver committed:
Fixed symbolic links for test files
